File: ios/Runner/Runner/Models/TrainingPlan.swift
import Foundation

struct TrainingPlan: Codable, Identifiable {
    let id: String
    let userId: String
    let title: String
    let description: String
    let duration: Int // weeks
    let difficulty: DifficultyLevel
    let goals: [String]
    let workouts: [Workout]
    let createdAt: Date
    let updatedAt: Date
}

struct Workout: Codable, Identifiable {
    let id: String
    let planId: String
    let title: String
    let description: String
    let type: WorkoutType
    let duration: Int // minutes
    let distance: Double? // km
    let intensity: IntensityLevel
    let scheduledDate: Date?
    let completed: Bool
    let instructions: [String]
    let fitFileUrl: String?
}

enum DifficultyLevel: String, CaseIterable, Codable {
    case beginner = "beginner"
    case intermediate = "intermediate"
    case advanced = "advanced"
    
    var displayName: String {
        switch self {
        case .beginner:
            return "Beginner"
        case .intermediate:
            return "Intermediate"
        case .advanced:
            return "Advanced"
        }
    }
}

enum WorkoutType: String, CaseIterable, Codable {
    case easy = "easy"
    case tempo = "tempo"
    case interval = "interval"
    case longRun = "long_run"
    case recovery = "recovery"
    case race = "race"
    case crossTraining = "cross_training"
    
    var displayName: String {
        switch self {
        case .easy:
            return "Easy Run"
        case .tempo:
            return "Tempo Run"
        case .interval:
            return "Interval Training"
        case .longRun:
            return "Long Run"
        case .recovery:
            return "Recovery Run"
        case .race:
            return "Race"
        case .crossTraining:
            return "Cross Training"
        }
    }
    
    var icon: String {
        switch self {
        case .easy:
            return "figure.walk"
        case .tempo:
            return "figure.run"
        case .interval:
            return "timer"
        case .longRun:
            return "figure.run.circle"
        case .recovery:
            return "heart.fill"
        case .race:
            return "flag.fill"
        case .crossTraining:
            return "dumbbell.fill"
        }
    }
}

enum IntensityLevel: String, CaseIterable, Codable {
    case low = "low"
    case moderate = "moderate"
    case high = "high"
    case maximum = "maximum"
    
    var displayName: String {
        switch self {
        case .low:
            return "Low"
        case .moderate:
            return "Moderate"
        case .high:
            return "High"
        case .maximum:
            return "Maximum"
        }
    }
    
    var color: String {
        switch self {
        case .low:
            return "green"
        case .moderate:
            return "yellow"
        case .high:
            return "orange"
        case .maximum:
            return "red"
        }
    }
}

File: backend/models/TrainingPlan.js
const mongoose = require('mongoose');

const workoutSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  type: {
    type: String,
    enum: ['easy', 'tempo', 'interval', 'long_run', 'recovery', 'race', 'cross_training'],
    required: true
  },
  duration: {
    type: Number, // minutes
    required: true
  },
  distance: {
    type: Number, // km
    required: false
  },
  intensity: {
    type: String,
    enum: ['low', 'moderate', 'high', 'maximum'],
    required: true
  },
  scheduledDate: {
    type: Date,
    required: false
  },
  completed: {
    type: Boolean,
    default: false
  },
  instructions: [{
    type: String
  }],
  fitFileUrl: {
    type: String,
    required: false
  }
}, {
  timestamps: true
});

const trainingPlanSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  duration: {
    type: Number, // weeks
    required: true
  },
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced'],
    required: true
  },
  goals: [{
    type: String
  }],
  workouts: [workoutSchema],
  inputData: {
    distance: Number,
    distanceUnit: {
      type: String,
      enum: ['km', 'miles']
    },
    time: {
      hours: Number,
      minutes: Number
    },
    maxDistance: Number,
    maxDistanceUnit: {
      type: String,
      enum: ['km', 'miles']
    }
  },
  aiGeneratedContent: {
    type: String, // Store the raw OpenAI response
    required: false
  }
}, {
  timestamps: true
});

// Index for efficient queries
trainingPlanSchema.index({ userId: 1, createdAt: -1 });
workoutSchema.index({ scheduledDate: 1 });

module.exports = mongoose.model('TrainingPlan', trainingPlanSchema);
